cache.go

```diff
@@ -1,91 +1,266 @@
+	evictions int64
+
+	items        map[interface{}]*list.Element
+	evictionList *list.List
+	mutex        sync.RWMutex
+}
+
+// New constructs an LRU Cache with the given Config object. Returns an error
+// given an invalid size.
+func New(config Config) (*Cache, error) {
+	if config.Size <= 0 {
+		return nil, errors.New("Must supply a positive config.Size")
 	}
 
-	c := &Cache{
-		lru:    l,
-		maxAge: maxAge,
+	if config.MaxAge < 0 {
+		return nil, errors.New("Must supply a zero or positive config.MaxAge")
 	}
 
-	return c, nil
+	cache := &Cache{
+		size:         config.Size,
+		maxAge:       config.MaxAge,
+		onEviction:   config.OnEviction,
+		onExpiration: config.OnExpiration,
+		items:        make(map[interface{}]*list.Element),
+		evictionList: list.New(),
+	}
+
+	return cache, nil
 }
 
-// Add adds an additional key/value pair to our cache.
-func (c *Cache) Add(key, value interface{}) bool {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+// Set updates a key:value pair in the cache. Returns true if an eviction
+// occurrred, and subsequently invokes the OnEviction callback.
+func (cache *Cache) Set(key, value interface{}) bool {
+	cache.mutex.Lock()
+	defer cache.mutex.Unlock()
 
-	return c.lru.Add(key, entry{
-		value:   value,
-		created: time.Now(),
-	})
+	cache.sets++
+	created := time.Now()
+
+	if element, ok := cache.items[key]; ok {
+		cache.evictionList.MoveToFront(element)
+		entry := element.Value.(*cacheEntry)
+		entry.value = value
+		entry.created = created
+		return false
+	}
+
+	entry := &cacheEntry{key, value, created}
+	element := cache.evictionList.PushFront(entry)
+	cache.items[key] = element
+
+	evict := cache.evictionList.Len() > cache.size
+	if evict {
+		cache.evictOldest()
+	}
+	return evict
 }
 
-// Get returns the value stored at `key`.
-//
-// The boolean value reports if the value was found.
-func (c *Cache) Get(key interface{}) (interface{}, bool) {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+// Get returns the value stored at `key`. The boolean value reports whether or
+// not the value was found. The OnExpiration callback is invoked if the value
+// had expired on access
+func (cache *Cache) Get(key interface{}) (interface{}, bool) {
+	cache.mutex.Lock()
+	defer cache.mutex.Unlock()
 
-	val, ok := c.lru.Get(key)
-	if !ok {
-		c.stats.Misses++
-		return nil, ok
+	cache.gets++
+
+	if element, ok := cache.items[key]; ok {
+		entry := element.Value.(*cacheEntry)
+		if cache.maxAge == 0 || time.Since(entry.created) <= cache.maxAge {
+			cache.evictionList.MoveToFront(element)
+			cache.hits++
+			return entry.value, true
+		}
+
+		// Entry expired
+		cache.deleteElement(element)
+		cache.misses++
+		if cache.onExpiration != nil {
+			cache.onExpiration(entry.key, entry.value)
+		}
+		return nil, false
 	}
 
-	e := val.(entry)
-	if time.Since(e.created) <= c.maxAge {
-		c.stats.Hits++
-		return e.value, true
+	cache.misses++
+	return nil, false
+}
+
+// HasKey returns whether or not the `key` is in the cache without updating
```

Comment:
should this be called `Has`? similar to `Get`, `Set` etc..

Reply:
I'm used to ruby's `has_key` to clarify it's not operating on values, but that works :)